### PR TITLE
Enforce typing in run_release.py and release.py

### LINKED ISSUE #114: Add a type checker
Might also be worth adding a type checker, which could have caught #107.

_Originally posted by @JelleZijlstra in https://github.com/python/release-tools/issues/109#issuecomment-2045309417_

### PR DESCRIPTION
**Pull request by ambv**
This depends on #151 landing first. (change base branch to `master` to land)

Closes #114.

### CODE REVIEW

File: release.py [hunk 1]
```diff
@@ -20,14 +20,110 @@
...
+    overload,
+)
 
 COMMASPACE = ", "
 SPACE = " "
 tag_cre = re.compile(r"(\d+)(?:\.(\d+)(?:\.(\d+))?)?(?:([ab]|rc)(\d+))?$")
 
 
-# Ideas stolen from Mailman's release script, Lib/tokens.py and welease
+class ReleaseShelf(Protocol):  # this will be easier to type in 3.13
```
Comment (ambv) [inline release.py:40]: This is like a `TypedDict` but for a pseudo-mapping not inheriting from a `dict`. The syntax is a bit horrible but it's per-key strongly typed.

And actually, this *won't* be easier in 3.13 :sob:
https://peps.python.org/pep-0705/#a-typedmapping-protocol-type

File: run_release.py [hunk 2]
```diff
@@ -652,25 +648,19 @@ def execute_command(command: str) -> None:
 
     # Docs
 
-    release_tag: release_mod.Tag = db["release"]
+    release_tag = db["release"]
     if release_tag.is_final or release_tag.is_release_candidate:
         source = f"/home/psf-users/{db['ssh_user']}/{db['release']}"
         destination = f"/srv/www.python.org/ftp/python/doc/{release_tag}"
 
-        def execute_command(command: str) -> None:
```
Comment (ambv) [inline run_release.py:660]: This is shadowing an identical function from L641

**Comment by hugovk** [comment]
#151 has landed, I resolved the conflicts.

File: release.py (hunk 1 above)
Comment (hugovk) [inline release.py:40]: Then remove the comment? 
```suggestion
class ReleaseShelf(Protocol):
```

**Review by hugovk — APPROVED** [review]

File: run_release.py [hunk 3]
```diff
@@ -179,15 +175,6 @@
 """
 
 
-@dataclass
-class Task:
```
Comment (sethmlarson) [inline run_release.py:183]: `Task` isn't used anywhere in `release.py` that I can see, is there a reason we moved it there?

**Review by sethmlarson — APPROVED** [review]
Only had one comment, otherwise LGTM!

File: run_release.py (hunk 3 above)
Comment (ambv) [reply run_release.py:183]: Yeah, it's used in typing of ReleaseShelf and it refers to ReleaseShelf itself. So it needs to be in the same spot.

**Pull request merged**